8331051: Add an `@since` checker test for `java.base` module

nizarbenalla opened the pull request:
This checker checks the values of the `@since` tag found in the documentation comment for an element against the release in which the element first appeared.

Real since value of an API element is computed as the oldest release in which the given API element was introduced. That is:
- for modules, classes and interfaces, the release in which the element with the given qualified name was introduced
- for constructors, the release in which the constructor with the given VM descriptor was introduced
- for methods and fields, the release in which the given method or field with the given VM descriptor became a member of its enclosing class or interface, whether direct or inherited

Effective since value of an API element is computed as follows:
- if the given element has a `@since` tag in its javadoc, it is used
- in all other cases, return the effective since value of the enclosing element

The since checker verifies that for every API element, the real since value and the effective since value are the same, and reports an error if they are not.

Preview method are handled as per JEP 12, if `@PreviewFeature` is used consistently going forward then the checker doesn't need to be updated with every release. The checker has explicit knowledge of preview elements that came before `JDK 14` because they weren't marked in a machine understandable way and preview elements that came before `JDK 17` that didn't have `@PreviewFeature`.

Important note : We only check code written since `JDK 9` as the releases used to determine the expected value of `@since` tags are taken from the historical data built into `javac` which only goes back that far

The intial comment at the beginning of `SinceChecker.java` holds more information into the program.

I already have filed issues and fixed some wrong tags like in #18640, #18032, #18030, #18055, #18373, #18954, #18972.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331051](https://bugs.openjdk.org/browse/JDK-8331051): Add an `@<!---->since` checker test for `java.base` module (**Sub-task** - P4)


### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) Review applies to [62a34064](https://git.openjdk.org/jdk/pull/18934/files/62a34064856dd19290218f4ece5150efb7d034d3)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18934/head:pull/18934` \
`$ git checkout pull/18934`

Update a local copy of the PR: \
`$ git checkout pull/18934` \
`$ git pull https://git.openjdk.org/jdk.git pull/18934/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18934`

View PR using the GUI difftool: \
`$ git pr show -t 18934`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18934.diff">https://git.openjdk.org/jdk/pull/18934.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18934#issuecomment-2082235634)

jonathan-gibbons commented on test/jdk/tools/sincechecker/testjavabase/SinceChecker.java:
```diff
@@ -0,0 +1,37 @@
+ * @run main SinceValidator java.base
+ */
+
+public class SinceChecker {
+}
```
You don't need this class, do you?
The comment should be enough to run the validator

nizarbenalla replied:
Thanks, didn't know that.

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+import java.util.stream.Collectors;
+
+/*
+The `@since` checker works as a two-step process:
```
Insert an initial sentence or paragraph, such as the following:

```
This checker checks the values of the `@since` tag found in the documentation comment for an element
against the release in which the element first appeared.

The source code containing the documentation comments is read from `src.zip` in the release of
JDK used to run the test.  The releases used to determine the expected value of `@since` tags
are taken from the historical data built into `javac`.
```

nizarbenalla replied:
Added in  32edb4d811978f87bd2f850c214191e8fc7b8c88

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+Special Handling for preview method:
+- When an element is still marked as preview, the `@since` should be the first JDK release where the element was added.
+- If the element is no longer marked as preview, the `@since` should be the first JDK release where it was no longer preview.
+
```
Add a comment about need for special treatment of early preview features.

nizarbenalla replied:
Thanks, added it

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        }
+    }
+
+    private void processModuleRecord(ModuleElement moduleElement, String releaseVersion, JavacTask ct) {
```
for this method, and similarly named methods, the use of `Record` seems confusing, if only because there do not seem to be any record classes being processed or analyzed.   Consider dropping `Record` or changing it to `Element`. Same for `analyzePackageRecord`, `analyzeClassRecord`.

Also, consider being consistent with `process` or analyze`.

nizarbenalla replied:
Fixed in 32edb4d811978f87bd2f850c214191e8fc7b8c88

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+
+    private void analyzeClassRecord(TypeElement te, String version, Types types, Elements elements) {
+        Set<Modifier> classModifiers = te.getModifiers();
+        if (!(classModifiers.contains(Modifier.PUBLIC) || classModifiers.contains(Modifier.PROTECTED))) {
```
Insert comment:
```
/// JDK documentation only contains public and protected declarations
```
which is 99% true. (The serialization page can contain `private` declarations...)

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        }
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(element -> element.getModifiers().contains(Modifier.PUBLIC) || element.getModifiers().contains(Modifier.PROTECTED))
```
Consider writing and using a predicate method:

```
boolean isDocumented(Element e) {
    var mods = e.getModifiers();
    return mods.contains(Modifier,.PUBLIC) || mods.contains(Modifier.PROTECTED);
}
```

You could then use that method in the obvious way for the class declaration(line 134) and in a lambda as follows:
```
    .filter(this::isDocumented)
```

nizarbenalla replied:
Added in 32edb4d811978f87bd2f850c214191e8fc7b8c88

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(element -> element.getModifiers().contains(Modifier.PUBLIC) || element.getModifiers().contains(Modifier.PROTECTED))
+                .filter(element -> element.getKind().isField()
+                        || element.getKind() == ElementKind.METHOD
+                        || element.getKind() == ElementKind.CONSTRUCTOR)
```
Consider writing and using another predicate method, 
```
    boolean isMember(Element e) {
        var k = e.getKind();
        return k.isField() || switch (k) {
            case CONSTRUCTOR, METHOD -> true;
            default -> false;
        };
     }
     
```

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        return prefix + suffix;
+    }
+
+    public static class IntroducedIn {
+        public String introducedPreview;
+        public String introducedStable;
+    }
```
Suggest move this nearer the map at the top of the class -- either move this class up, or those members down.

nizarbenalla replied:
Thanks, moved it up

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+
+public class SinceChecker {
+    private final Map<String, Set<String>> LEGACY_PREVIEW_METHODS = new HashMap<>();
+    private final Map<String, IntroducedIn> classDictionary = new HashMap<>();
```
See comment lower down about `class IntroducedIn`

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+            Path testJdk = Paths.get(System.getProperty("test.jdk"));
+            srcZip = testJdk.getParent().resolve("images").resolve("jdk").resolve("lib").resolve("src.zip");
+        }
+        File f = new File(srcZip.toUri());
+        if (!f.exists() && !f.isDirectory()) {
+            throw new SkippedException("Skipping Test because src.zip wasn't found");
+        }
```
For modern/new code I would recommend using `java.nio.file.Path` instead of `java.io.File`.

I see you actually have a `Path` in `srcZip`, so use `Files.exists(srcZip)` and `Files.isDirectory(srcZip)`

nizarbenalla replied:
Fixed in 32edb4d811978f87bd2f850c214191e8fc7b8c88

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+                        wrongTagsList.add("Initiating javadocHelperFailed " + e.getMessage());
+                    }
+                    if (!wrongTagsList.isEmpty()) {
+                        throw new Exception(wrongTagsList.toString());
```
This looks like it might be the exception to end the test, right, and the wrongTags could be long.
I'd suggest writing a heading to `System.err`, then`wrongTagsList.forEach(System.err::println);` and then throw the exception with a summary string, like `throw new Exception(wrongTagsList.size() + " problems found");

nizarbenalla replied:
Done. Added in 32edb4d811978f87bd2f850c214191e8fc7b8c88

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+
+    private void processModuleCheck(ModuleElement moduleElement, JavacTask ct, Path packagePath, EffectiveSourceSinceHelper javadocHelper) {
+        if (moduleElement == null) {
+            wrongTagsList.add("Module element: was null because `elements.getModuleElement(moduleName)` returns null." +
```
This doesn't look like a "Wrong tag" so much as a general error message.

Consider this as a different paradigm -- instead of saving up strings in `wrongTagsList`, consider having a method to report (and count) error messages as you find them:

```
    private int errorCount;
    void error(String message) {
        System.err.println(message);
        errorCount++;
   }
```

then at the end of `testThisModule`, just check if `errorCount>0` to decide whether to throw the exception.

Note, in this model, you need not have a `\n` at the end of every string argument, the way you do for add strings added into `wrongTagsList`.

nizarbenalla replied:
Changed it and removed '\n'

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+                && LEGACY_PREVIEW_METHODS.get(currentVersion).contains(uniqueId);
+    }
+
+    private void testThisModule(String moduleName) throws Exception {
```
The word `this` is a bit distracting/unnecessary here.  It's enough to say `testModule(String moduleName)` although `checkModule(String moduleName)` would be even better

nizarbenalla replied:
Thanks, renamed it to `CheckModule`

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        String packageTopVersion = null;
+        if (Files.exists(pkgInfo)) {
+            try {
+                byte[] packageAsBytes = Files.readAllBytes(pkgInfo);
+                String packageContent = new String(packageAsBytes, StandardCharsets.UTF_8);
```
Easier to just use `Files.readString`

nizarbenalla replied:
Fixed it, thanks

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+    }
+
+    private void analyzeClassCheck(TypeElement te, String version, EffectiveSourceSinceHelper javadocHelper,
+                                   Types types, Elements elementUtils) {
+        String currentjdkVersion = String.valueOf(Runtime.version().feature());
+        Set<Modifier> classModifiers = te.getModifiers();
+        if (!(classModifiers.contains(Modifier.PUBLIC) || classModifiers.contains(Modifier.PROTECTED))) {
+            return;
+        }
+        checkElement(te.getEnclosingElement(), te, types, javadocHelper, version, elementUtils);
+        te.getEnclosedElements().stream().filter(element -> element.getModifiers().contains(Modifier.PUBLIC)
+                        || element.getModifiers().contains(Modifier.PROTECTED))
+                .filter(element -> element.getKind().isField()
+                        || element.getKind() == ElementKind.METHOD
+                        || element.getKind() == ElementKind.CONSTRUCTOR)
+                .forEach(element -> checkElement(te, element, types, javadocHelper, version, elementUtils));
+        te.getEnclosedElements().stream()
+                .filter(element -> element.getKind().isDeclaredType())
+                .map(TypeElement.class::cast)
+                .forEach(nestedClass -> analyzeClassCheck(nestedClass, currentjdkVersion, javadocHelper, types, elementUtils));
+    }
```
more possible applications of those utility predicate methods I mentioned earlier.

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+                    mappedVersion.introducedPreview :
+                    mappedVersion.introducedStable;
+        } catch (Exception e) {
+            wrongTagsList.add("For element " + element + "mappedVersion" + mappedVersion + " is null" + e + "\n");
```
Put a space e or colon between `is null` and `e`

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        ));
+    }
+
+    private final class EffectiveSourceSinceHelper implements AutoCloseable {
```
Consider putting a doc comment describing the purpose of this class. The name  hints at something useful, but is not specific enough by itself.

nizarbenalla replied:
Added it, thanks

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+            fm.close();
+        }
+
+        private static final class PatchModuleFileManager
```
provide a short doc comment describing the purpose/use of this class

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+                try {
+                    fm.close();
+                } catch (IOException closeEx) {
+                }
```
Consider using `ex.addSuppressed(closeEx);` instead of just dropping the exception

nizarbenalla replied:
Fixed in 32edb4d811978f87bd2f850c214191e8fc7b8c88

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        String message;
+        if (mappedVersion.equals("9")) {
+            message = "For " + elementSimpleName +
+                    " Wrong `@since` version " + sinceVersion + " But the element exists before JDK 10\n";
```
lower case "but"

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+
+    private void checkEquals(String sinceVersion, String mappedVersion, String id) {
+        if (sinceVersion == null || mappedVersion == null) {
+            wrongTagsList.add("For " + id + " NULL value for either mapped or real `@since` . mapped version is="
```
Here and elsewhere, instead of `"For " + name + ....`  consider using the format `name + ":  " + ...` for the format of the messages

nizarbenalla replied:
Fixed it.

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,769 @@
+        public String introducedStable;
+    }
+
+    //these were preview in before the introduction of the @PreviewFeature
```
Just curious: where do you find this information?

nizarbenalla replied:
Used a script to get all the elements with `@Deprecated(forRemoval=true, since=...)`, `@jdk.internal.preview`, `{@preview Associated with ....` in their javadoc.
And then had to cross reference with a document that had all the ids

jonathan-gibbons replied:
Thanks for the update

jonathan-gibbons reviewed (commented):
Various comments inline.  Overall, a great start.
Consider adding more explanatory comments for the bigger items, like classes and long/important methods. Imagine a future-person looking over your shoulder, asking what the more important items do or are for.

jddarcy commented:
If not already planned, I recommend a much broader discussion/announcement of this test before it goes back. For example, a message to jdk-dev stating "The value of `@since` tags will be tested now by a tier $N test. This is how you run the test before a push. Here are rules for some tricky cases like introducing a covariant override." etc.

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+            URI uri = URI.create("jar:" + srcZip.toUri());
+            try (FileSystem zipFO = FileSystems.newFileSystem(uri, Collections.emptyMap())) {
+                Path root = zipFO.getRootDirectories().iterator().next();
+                Path packagePath = root.resolve(moduleName);
```
Here, and infuses of this name, including in method signatures, it is strange to see the variable named `packagePath` since it is the path of the unnamed package in the module.

`modulePath` would likely be a better name, unless you are worried about possible confusion with the `--module-path` option, in which case maybe `moduleDir` or `moduleDirectory` would work.

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+        }
+    }
+
+    private void processModuleCheck(ModuleElement moduleElement, JavacTask ct, Path packagePath, EffectiveSourceSinceHelper javadocHelper) {
```
see comment line 223, about use of `packagePath`

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+        }
+    }
+
+    private String getModuleVersionFromFile(Path packagePath) {
```
see comment line 223, about `packagePath`

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+        return version;
+    }
+
+    private String getPackageVersionFromFile(Path packagePath, ModuleElement.ExportsDirective ed) {
```
see comment line 223, about `packagePath`

jonathan-gibbons commented on test/jdk/TEST.groups:
```diff
@@ -661,4 +662,8 @@ jdk_containers_extended = \
+
+# Set of tests for `@since` checks in source code documentation
+jdk_since_check = \
+   tools/sincechecker/testjavabase/CheckSince_javaBase.java
```
Generally, the symbol of a red line in a red circle is indicative of a missing final newline at the end of the file.
It is recommend that files end with exactly one newline character.

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+note on rules for Real and effective `@since:
+
+Real since value of an API element is computed as the oldest release in which the given API element was introduced. That is:
+- for modules, classes and interfaces, the release in which the element with the given qualified name was introduced
```
What about packages?  packages should be in this list as well.

nizarbenalla replied:
Small oversight, will fix

nizarbenalla commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+
+
+/*
+- This checker checks the values of the `@since` tag found in the documentation comment for an element against the release in which the element first appeared.
+- The source code containing the documentation comments is read from `src.zip` in the release of JDK used to run the test.
+- The releases used to determine the expected value of `@since` tags are taken from the historical data built into `javac`.
```
My idea was that, even when comparing files in a merge conflit, there wouldn't be problems here. 

But I will fix, some lines in the code are a bit long as well

hns commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+
+
+/*
+- This checker checks the values of the `@since` tag found in the documentation comment for an element against the release in which the element first appeared.
```
In addition to the long lines mentioned by Jon which make the comments hard to read, I find it strange that every sentence is formatted as a list item with a leading dash. I think it's ok when describing different parts/steps of the algorithm, but at least the first sentence in the comment should not be a list item IMO.

nizarbenalla replied:
Fixed, thanks.

hns commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+
+    public static void main(String[] args) throws Exception {
+        if (args.length == 0) {
+            throw new SkippedException("Test module not specified");
```
I don't think `SkippedException` is the right exception to throw here, since invoking the method with a missing argument is probably a configuration error that shouldn't be ignored. Maybe `IllegalArgumentException` or just `RuntimeException`?

nizarbenalla replied:
Fixed it, now throwing `IllegalArgumentException`.

hns commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+                .forEach(nestedClass -> processClassElement(nestedClass, version, types, elements));
+    }
+
+    public void persistElement(Element explicitOwner, Element element, Types types, String version) {
```
I'm not sure `persistElement` is a good name for this, it sounds like the element is being stored permanently (beyond the runtime of this program). I can't think of a really good name, but I think `processElement` would be ok.

hns commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+            Path testJdk = Paths.get(System.getProperty("test.jdk"));
+            srcZip = testJdk.getParent().resolve("images").resolve("jdk").resolve("lib").resolve("src.zip");
+        }
+        if (!Files.exists(srcZip) && !Files.isDirectory(srcZip)) {
```
This condition looks wrong. If `exists` returns false, it also implies that `isDirectory` returns false. I think there's no need to check for a (not-)directory here.

nizarbenalla replied:
Fixed in [3f226ef](https://github.com/openjdk/jdk/pull/18934/commits/3f226ef9134b71a1b63b6562b8381be909c30343), I now only check if the file is readable.

nizarbenalla commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,828 @@
+        Path pkgInfo = packagePath.resolve(ed.getPackage()
+                        .getQualifiedName()
+                        .toString()
+                        .replaceAll("\\.", "/")
```
Replaced with `.replace(".", File.separator)`

nizarbenalla commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,842 @@
+
+    private void processModuleElement(ModuleElement moduleElement, String releaseVersion, JavacTask ct) {
+        for (ModuleElement.ExportsDirective ed : ElementFilter.exportsIn(moduleElement.getDirectives())) {
+            processElement(moduleElement, moduleElement, ct.getTypes(), releaseVersion);
```
For consideration for tomorrow: move this one line up - to map the module id even if the module has no export directives

lahodaj commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,878 @@
+            return;
+        }
+        checkElement(te.getEnclosingElement(), te, types, javadocHelper, version, elementUtils);
+        if( te.getKind() == ElementKind.RECORD){
```
This seems a bit too broad. Ignoring all members of a record type may lead to missed members. `Elements.getOrigin` may be unusable here, but I wonder what exactly is the problem which this condition is trying to solve?

nizarbenalla replied:
I changed my approach in [ac4df85](https://github.com/openjdk/jdk/pull/18934/commits/ac4df85615557306da606a712ee02e7155e5412f). I beleieve it is a bit clearer now. I am only skipping some very common methods (`toString`, `equals` and `hashCode`)

Previously, I failed to get the effective `@since` of some common methods eclosed within a record.
I will use this as an example
https://github.com/openjdk/jdk/blob/b92bd671835c37cff58e2cdcecd0fe4277557d7f/src/jdk.jshell/share/classes/jdk/jshell/execution/JdiDefaultExecutionControl.java#L346

`te.getEnclosedElements()` on this record returns the following (I truncated `jdk.jshell.execution.JdiDefaultExecutionControl.JdiStarter.TargetDescription` into `TargetDescription`)
```
method: void TargetDescription.<init>(com.sun.jdi.VirtualMachine,java.lang.Process) 
method: void TargetDescription.wait(long,int) 
method: java.lang.Process TargetDescription.process() 
method: com.sun.jdi.VirtualMachine TargetDescription.vm() 
method: java.lang.String TargetDescription.toString() 
method: int TargetDescription.hashCode() 
method: java.lang.Object TargetDescription.clone() 
method: java.lang.Class TargetDescription.getClass() 
method: void TargetDescription.finalize() 
method: void TargetDescription.notify() 
method: void TargetDescription.wait(long) 
method: boolean TargetDescription.equals(java.lang.Object) 
method: void TargetDescription.notifyAll() 
method: void TargetDescription.wait() 
```

Getting the effective `@since` for these 5 methods in particular would fail
```
method: java.lang.String TargetDescription.toString():
method: int TargetDescription.hashCode(): 
method: boolean TargetDescription.equals(java.lang.Object):
method: com.sun.jdi.VirtualMachine TargetDescription.vm(): 
method: java.lang.Process TargetDescription.process():
```
It seemed to me that these methods would be added in the same version as the record, so I skipped them all.
I didn't think of a user adding a static variable in a later release and that it was too broad, my bad.

lahodaj reviewed (commented):
Overall seems pretty good, great job!

A few comments inline, to make the code cleaner.

Thanks for working on this!

nizarbenalla commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,878 @@
+            } catch (IOException e) {
+                error("module-info.java not found or couldn't be opened AND this module has no unqualified exports");
+            } catch (NullPointerException e) {
+                error("module-info.java does not contain an `@since`");
```
Fixed in [e82dfbf](https://github.com/openjdk/jdk/pull/18934/commits/e82dfbf0f3df1dfcb1e482aac09fb34e39af9be0), I tried not catching NPE before but found the extra checks a bit ugly.

nizarbenalla commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,878 @@
+            el = el.getEnclosingElement();
+        }
+
+        return LEGACY_PREVIEW_METHODS.containsKey(currentVersion)
```
Fixed, thanks.

lahodaj commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -342,18 +342,38 @@ private void analyzePackageCheck(PackageElement pe, JavacTask ct, EffectiveSourc
+            return true;
+        }
+        String uniqueId = getElementName(te, element, types);
+        boolean isCommonMethod = uniqueId.endsWith("toString()") ||
```
Nit: `.toString()` instead of `toString()`?

lahodaj commented:
/reviewers 2

lahodaj reviewed (approved):
Looks good to me, thanks for doing this!

nizarbenalla commented:
The checker's report has been improved, the result after running it on `java.base` is the following
```
        SinceChecker java.base
STDERR:
src/java.base/java/io/PrintStream.java:685 method: void java.io.PrintStream.write(byte[]): `@since` version is 14 but the element exists before JDK 10
src/java.base/java/io/FileInputStream.java:374 method: byte[] java.io.FileInputStream.readNBytes(int): `@since` version: 9; found: 11
src/java.base/java/lang/classfile/ClassSignature.java:45 method: java.lang.classfile.Signature.ClassTypeSig java.lang.classfile.ClassSignature.superclassSignature(): `@since` version: 22; found: 23
src/java.base/java/lang/classfile/ClassReader.java:131 method: java.lang.classfile.constantpool.PoolEntry java.lang.classfile.ClassReader.readEntryOrNull(int,java.lang.Class): `@since` version: 24; found: 23
src/java.base/java/lang/classfile/constantpool/ConstantPool.java:74 method: java.lang.classfile.constantpool.PoolEntry java.lang.classfile.constantpool.ConstantPool.entryByIndex(int,java.lang.Class): `@since` version: 24; found: 23
src/java.base/java/lang/constant/ClassDesc.java:378 method: java.lang.Class java.lang.constant.ClassDesc.resolveConstantDesc(java.lang.invoke.MethodHandles.Lookup): `@since` version: 12; found: 21
src/java.base/java/lang/constant/MethodTypeDesc.java:226 method: java.lang.invoke.MethodType java.lang.constant.MethodTypeDesc.resolveConstantDesc(java.lang.invoke.MethodHandles.Lookup): `@since` version: 12; found: 21
src/java.base/java/lang/constant/MethodHandleDesc.java:210 method: java.lang.invoke.MethodHandle java.lang.constant.MethodHandleDesc.resolveConstantDesc(java.lang.invoke.MethodHandles.Lookup): `@since` version: 12; found: 21
src/java.base/java/lang/foreign/MemorySegment.java:622 method: long java.lang.foreign.MemorySegment.maxByteAlignment(): `@since` version: 22; found: 23
src/java.base/java/lang/invoke/MethodHandles.java:7923 method: java.lang.invoke.MethodHandle java.lang.invoke.MethodHandles.tableSwitch(java.lang.invoke.MethodHandle,java.lang.invoke.MethodHandle[]): `@since` version: 9; found: 17
src/java.base/java/lang/ref/Reference.java:527 method: java.lang.Object java.lang.ref.Reference.clone(): `@since` version is 11 but the element exists before JDK 10
src/java.base/java/lang/reflect/MalformedParameterizedTypeException.java:54 method: void java.lang.reflect.MalformedParameterizedTypeException.<init>(java.lang.String): `@since` version: 9; found: 10
src/java.base/java/nio/MappedByteBuffer.java:401 method: java.nio.MappedByteBuffer java.nio.MappedByteBuffer.slice(): `@since` version: 9; found: 17
src/java.base/java/nio/MappedByteBuffer.java:414 method: java.nio.MappedByteBuffer java.nio.MappedByteBuffer.slice(int,int): `@since` version: 9; found: 17
src/java.base/java/nio/MappedByteBuffer.java:420 method: java.nio.MappedByteBuffer java.nio.MappedByteBuffer.duplicate(): `@since` version: 9; found: 17
src/java.base/java/nio/MappedByteBuffer.java:427 method: java.nio.MappedByteBuffer java.nio.MappedByteBuffer.compact(): `@since` version: 9; found: 17
src/java.base/java/text/ChoiceFormat.java:589 method: boolean java.text.ChoiceFormat.isStrict(): `@since` version: 9; found: 23
src/java.base/java/text/ChoiceFormat.java:595 method: void java.text.ChoiceFormat.setStrict(boolean): `@since` version: 9; found: 23
src/java.base/java/util/Properties.java:189 method: void java.util.Properties.<init>(int): `@since` version: 9; found: 10
src/java.base/java/util/concurrent/FutureTask.java:208 method: java.lang.Object java.util.concurrent.FutureTask.resultNow(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/FutureTask.java:224 method: java.lang.Throwable java.util.concurrent.FutureTask.exceptionNow(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/FutureTask.java:239 method: java.util.concurrent.Future.State java.util.concurrent.FutureTask.state(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/ThreadLocalRandom.java:504 method: float java.util.concurrent.ThreadLocalRandom.nextFloat(float): `@since` version: 9; found: 17
src/java.base/java/util/concurrent/ThreadLocalRandom.java:514 method: float java.util.concurrent.ThreadLocalRandom.nextFloat(float,float): `@since` version: 9; found: 17
src/java.base/java/util/concurrent/ForkJoinTask.java:890 method: java.util.concurrent.Future.State java.util.concurrent.ForkJoinTask.state(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/ForkJoinTask.java:899 method: java.lang.Object java.util.concurrent.ForkJoinTask.resultNow(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/ForkJoinTask.java:913 method: java.lang.Throwable java.util.concurrent.ForkJoinTask.exceptionNow(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/DelayQueue.java:331 method: java.util.concurrent.Delayed java.util.concurrent.DelayQueue.remove(): `@since` version: 9; found: 21
src/java.base/java/util/concurrent/CompletableFuture.java:2137 method: java.lang.Object java.util.concurrent.CompletableFuture.resultNow(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/CompletableFuture.java:2152 method: java.lang.Throwable java.util.concurrent.CompletableFuture.exceptionNow(): `@since` version: 9; found: 19
src/java.base/java/util/concurrent/CompletableFuture.java:2399 method: java.util.concurrent.CompletableFuture java.util.concurrent.CompletableFuture.exceptionallyAsync(java.util.function.Function): `@since` version: 9; found: 12
src/java.base/java/util/concurrent/CompletableFuture.java:2404 method: java.util.concurrent.CompletableFuture java.util.concurrent.CompletableFuture.exceptionallyAsync(java.util.function.Function,java.util.concurrent.Executor): `@since` version: 9; found: 12
src/java.base/java/util/concurrent/CompletableFuture.java:2409 method: java.util.concurrent.CompletableFuture java.util.concurrent.CompletableFuture.exceptionallyCompose(java.util.function.Function): `@since` version: 9; found: 12
src/java.base/java/util/concurrent/CompletableFuture.java:2414 method: java.util.concurrent.CompletableFuture java.util.concurrent.CompletableFuture.exceptionallyComposeAsync(java.util.function.Function): `@since` version: 9; found: 12
src/java.base/java/util/concurrent/CompletableFuture.java:2419 method: java.util.concurrent.CompletableFuture java.util.concurrent.CompletableFuture.exceptionallyComposeAsync(java.util.function.Function,java.util.concurrent.Executor): `@since` version: 9; found: 12
src/java.base/java/util/concurrent/CompletableFuture.java:2544 method: java.util.concurrent.Future.State java.util.concurrent.CompletableFuture.state(): `@since` version: 9; found: 19
src/java.base/java/util/zip/Deflater.java:340 method: void java.util.zip.Deflater.setDictionary(java.nio.ByteBuffer): `@since` version: 9; found: 11
java.lang.Exception: The `@since` checker found 37 problems
        at SinceChecker.checkModule(SinceChecker.java:262)
        at SinceChecker.main(SinceChecker.java:123)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
        at java.base/java.lang.Thread.run(Thread.java:1575)

JavaTest Message: Test threw exception: java.lang.Exception
JavaTest Message: shutting down test


TEST RESULT: Failed. Execution failed: `main' threw exception: java.lang.Exception: The `@since` checker found 37 problems
--------------------------------------------------
Test results: failed: 1
```

nizarbenalla commented:
While this might seem like a lot, if a developer runs it before pushing, they should only encounter a few reports at most.

nizarbenalla commented:
I have a couple of small changes after the first approval, mainly improving the error messages. And avoiding false reports in case a file doesn't exist in `src.zip`

nizarbenalla commented:
Keep alive, this will need a re-review eventually because of the new rules.

nizarbenalla commented:
Currently running tier1,  I added the option to exclude certain packages in an attempt to integrate this faster while we fix other areas.

nizarbenalla commented:
I performed a merge and a tiny change to the error message by the test.
The major changes since the last review are that I now skip packages if they are not found, and I added an exclude list to skip certain packages if we wish to do so.

Here is an example of the error output of the tool 
```
STDERR:
src/java.base/java/lang/classfile/TypeKind.java:73 field: java.lang.classfile.TypeKind:BOOLEAN: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:80 field: java.lang.classfile.TypeKind:BYTE: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:87 field: java.lang.classfile.TypeKind:CHAR: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:94 field: java.lang.classfile.TypeKind:SHORT: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:99 field: java.lang.classfile.TypeKind:INT: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:103 field: java.lang.classfile.TypeKind:LONG: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:107 field: java.lang.classfile.TypeKind:FLOAT: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:111 field: java.lang.classfile.TypeKind:DOUBLE: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:117 field: java.lang.classfile.TypeKind:REFERENCE: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:125 field: java.lang.classfile.TypeKind:VOID: `@since` version: 22; should be: 24
src/java.base/java/lang/classfile/TypeKind.java:143 method: java.lang.constant.ClassDesc java.lang.classfile.TypeKind.upperBound(): `@since` version: 22; should be: 24
java.lang.Exception: The `@since` checker found 64 problems
        at SinceChecker.checkModule(SinceChecker.java:265)
        at SinceChecker.main(SinceChecker.java:126)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:573)
        at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
        at java.base/java.lang.Thread.run(Thread.java:1576)

```

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,926 @@
+
+            String version = String.valueOf(i);
+            Elements elements = ct.getElements();
+            elements.getModuleElement("java.base"); // forces module graph to be instantiated
```
@lahodaj I've always considered this kind of line to be a workaround; it would be nice if we could have an RFE so this kind of line was unnecessary

lahodaj replied:
Thanks. I've filled:
https://bugs.openjdk.org/browse/JDK-8341342

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,926 @@
+                version = effectiveSinceVersion(element.getEnclosingElement().getEnclosingElement(), element.getEnclosingElement(), types, elementUtils);
+            }
+            if (version == null) {
+                //may be null for private elements
```
empty statement for `if`...

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,926 @@
+import com.sun.tools.javac.util.Pair;
+import jtreg.SkippedException;
+
+/*
```
Good comment.

jonathan-gibbons commented on test/jdk/tools/sincechecker/SinceChecker.java:
```diff
@@ -0,0 +1,926 @@
+    private int errorCount = 0;
+
+    // packages to skip during the test
+    private static final Set<String> EXCLUDE_LIST = Set.of(
+            "java.lang.classfile"
+    );
```
This should not really be here, in the general code for a SinceChecker for all modules.

It would be better if this info was provided in command-line options, given when the test is run from the module-specific test (i.e. `CheckSince_javaBase.java`). 

This means you will have to write some additional simple option decoding in the `main` method, starting line 21.

jonathan-gibbons commented on test/jdk/tools/sincechecker/modules/java_base/CheckSince_javaBase.java:
```diff
@@ -0,0 +1,34 @@
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ *          jdk.compiler/com.sun.tools.javac.util
+ *          jdk.compiler/com.sun.tools.javac.code
+ * @run main SinceChecker java.base
```
See the comment about init-ing the EXCLUDE_LIST in the main SinceChecker, at about line 112.

You are already paying a module name as an option to the main test. Maybe the command here on line 33 could be something like
```
@run main SinceChecker java.base --exclude java.lang.classfile
```
where the argument after `--exclude` is (logically) a comma-separated list of items to exclude. You could also do a whitespace-separated list up to the next option (i.e. item beginning with `-`) or end of the list. In other words make it future-proof against adding more options in future.

jonathan-gibbons reviewed (changes requested):
<!-- Start of TOU acceptance message header -->
Hi @jonathan-gibbons, thanks for making a comment in an OpenJDK project!

All comments and discussions in the OpenJDK Community must be made available under the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/). If you already are an OpenJDK [Author](https://openjdk.java.net/bylaws#author), [Committer](https://openjdk.java.net/bylaws#committer) or [Reviewer](https://openjdk.java.net/bylaws#reviewer), please click [here](https://bugs.openjdk.java.net/secure/CreateIssue.jspa?pid=11300&issuetype=1) to open a new issue so that we can record that fact. Please Use "Add GitHub user jonathan-gibbons" for the summary.

If you are not an OpenJDK Author, Committer or Reviewer, simply check the box below to accept the OpenJDK Terms of Use for your comments.
<!-- End of TOU acceptance message header -->
 - [ ]  I agree to the [OpenJDK Terms of Use](https://openjdk.java.net/legal/tou/) for all comments I make in a project in the [OpenJDK GitHub organization](https://github.com/openjdk).
<!-- Start of TOU acceptance message footer -->
Your comment will be automatically restored once you have accepted the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/).
<!-- End of TOU acceptance message footer -->
<!-- Original message to be restored:
Generally+very+good.+It%27s+getting+to+be+a+really+nice+piece+of+functionality.%0D%0A%0D%0AI+added+one+suggestion+for+a+minor+improvement%2C+about+setting+up+the+%60EXCLUDE_LIST+in+a+more+module-specific+way.++The+intent+is+that+if+we+need+to+make+module-specific+changes+to+the+exclude+list+in+times+going+forward%2C+we+should+do+that+in+the+module-specific+test+invocation%2C+and+not+in+the+main+module-independent+code.
-->

jonathan-gibbons commented:
For discussion:

I see that in GitHub Actions, the test passes on all platforms. (That's good!). But it's not clear to me that it _needs_ to run on all platforms. Generally, the API should be the same on all platforms, right (After all, we only generate one set of API documentation.) 

So, the question becomes, can a test determine if it is being run in a multi-platform context?  We presumably don't want to stop the test running whenever a user runs it locally on their platform of choice, but equally, if the test is being run on many platforms, that's an unnecessary use/waste of resources.  But, I guess we run a lot of tests that are essentially platform-independent on all platforms, although I guess there're are also testing the product running on top of the JVM, whereas here. we're primarily just checking the source.

Anyway, this is really just an abstract academic discussion, not a blocker in any way for the PR.

lahodaj reviewed (approved):
Looks reasonable to me.

lahodaj reviewed (approved):
Still looks good to me.

nizarbenalla commented:
Can we maybe set the reviewer count to 1? This needs an additional review and the last change was small.

lahodaj commented:
/reviewers 1

nizarbenalla commented:
Thank you for the reviews; all. And the help to getting this work integrated.

/integrate
